server/graphql/v2/mutation/PaymentMethodMutations.js

```diff
@@ -78,8 +80,35 @@ const addCreditCard = {
       },
     };
 
+    // Check if the credit card is already saved
+    const oldPaymentMethod = await models.PaymentMethod.findOne({
+      where: {
+        CollectiveId: collective.id,
+        service: 'stripe',
+        type: 'creditcard',
+        saved: true,
+        data: {
+          fingerprint: token.card.fingerprint,
+        },
+      },
+    });
+    if (oldPaymentMethod) {
+      await oldPaymentMethod.destroy();
+    }
+
     let pm = await models.PaymentMethod.create(newPaymentMethodData);
 
+    // Update all existing orders with the new payment method
+    await models.Order.update(
+      { PaymentMethodId: pm.id },
+      {
+        where: {
+          PaymentMethodId: oldPaymentMethod.id,
+          status: { [Op.notIn]: ['PAID', 'CANCELLED', 'EXPIRED', 'DISPUTED', 'REFUNDED'] },
+        },
+      },
+    );
+
```

Comment:
The order here is not correct:
- Delete old PM
- Create new PM
- Update orders

Creating the new PM can fail, and we would end up in a broken state where the old PM is deleted but the orders are not updated with a new one. Or concurrency issues: recurring contributions are being processed, I'm updating my payment method and the order still holds a reference to the deleted PM.

**Solution:** use a sequelize transaction.

Reply:
And why not re-use the existing PM if it already exists, rather than deleting it? Like:


```es6
let paymentMethod = await findExistingCreditCard(fingerprint);
if (!paymentMethod || !checkPermissions(paymentMethod)) {
  paymentMethod = await models.PaymentMethod.create(newPaymentMethodData); 
}

Reply:
@Betree : Thanks. I've added the sequelize transactions, and added the shared helpers. 

> And why not re-use the existing PM if it already exists, rather than deleting it? Like:
> 
> ```js
> let paymentMethod = await findExistingCreditCard(fingerprint);
> if (!paymentMethod || !checkPermissions(paymentMethod)) {
>   paymentMethod = await models.PaymentMethod.create(newPaymentMethodData); 
> }
> ```

This one isn't possible I think since if we try to use the existing PM to update the card then Stripe will give an error such as, 

>The token tok_1MhSiXDjPFcHOcTmpnbDMKmh has already been used. Each token can only be used once to create a source.

because the token that is returned by Stripe should be different given we changed the card details.